plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -142,6 +147,20 @@ termEvaluationOrder semvar varStrictness = goTerm
 
         -- If we can view as a builtin application, then handle that specially
         (splitApplication -> (Builtin a fun, args)) -> goBuiltinApp a fun args
+        -- If we can view as a constructor application, then handle that specially.
+        -- Constructor applications are always pure: if under-applied they don't
+        -- reduce; if fully-applied they are pure; if over-applied it's going to be
+        -- a type error since they never return a function. So we can ignore the arity
+        -- in this case!
+        (splitApplication -> (h@(Var _ n), args))
+          | Just (DatatypeConstructor{}) <- lookupVarInfo n vinfo ->
+            evalThis (EvalTerm Pure MaybeWork h)
```

Comment:
Why `MaybeWork`? Does a constructor application do any non-trivial work?

Reply:
Unclear - I'm generally unclear on what "work free" should actually mean. At the moment the only things that are "work free" are immediately returning a value, which a constructor application certainly does not do.

Reply:
Oh, I thought `h` is the whole constructor application. If it is the head - the constructor itself, then shouldn't it always be WorkFree?

Also, shouldn't there be an `evalThis (EvalTerm Pure MaybeWork t)` after `goAppCtx args`? Because `t` is the constructor application, and `t` may not be workfree like you said.

Btw, `Unknown` seems redundant - can we use `EvalTerm Impure MaybeWork` in place of `Unknown`?

Reply:
> Also, shouldn't there be an evalThis (EvalTerm Pure MaybeWork t) after goAppCtx args?

Nvm about this, I see you just added it!

Reply:
> If it is the head - the constructor itself, then shouldn't it always be WorkFree?

Unsure: this is the same as we do for variables. Is accessing a variable work-free? You have to look it up... maybe that counts as work? IDK, I just stuck to what was there.

> Btw, Unknown seems redundant - can we use EvalTerm Impure MaybeWork in place of Unknown?

Yes, that would be a slightly different way of looking at it. Unknown has two purposes: it tells us that the evaluation goes somewhere unknown, which lets us distinguish terms for which we can see the whole evaluation order with ones where we actually just don't know what happens. If we only care about purity/work, then we can approximate that with "something that is maybe impure and does work", but I think it's kind of nice to say that it's an unknown jump. Also, it's what we use to terminate the traversal: we don't do any more stuff after we hit an Unknown, which is what stops us traversing the whole term in that case.